### ISSUE TITLE
Consider only synthesizing spaces when space glyph has no contours

### ISSUE THREAD

**Issue by drott** [harfbuzz/harfbuzz#2885]
In [crbug.com/1181275](https://bugs.chromium.org/p/chromium/issues/detail?id=1181275) the reporter observes that HarfBuzz synthesises a space glyph in debugging font that visualizes spaces with a dot.

This has undesirable effects for this font and debugging purpose when U+200B (which is not in the `cmap` of this font) gets synthesised from the space glyph, and the dot gets placed into the shape result with zero advance.

```
hb-view fontoxmlwhitespace.otf --unicodes=41,200b,41
```

![zero_advance_dot](https://user-images.githubusercontent.com/1021784/109826395-6ab79000-7c43-11eb-9e97-417f59e44000.png)

While whitespace synthetization is usually desirable to avoid expensive fallback for no actual glyph to be displayed, it may be worthwhile considering to disable it when the space glyph has contours? On the other hand, this is a special-purpose edge case, and there is a possibility for a font side workaround to map U+200B to an actual whitespace glyph.

CC @kojiishi

**Comment by jfkthame** [comment]
Maybe if `space` has contours, harfbuzz should consider checking for a NULL glyph at index 1 and/or by looking at how U+0000 or U+0008 is mapped. The [Apple TrueType spec](https://developer.apple.com/fonts/TrueType-Reference-Manual/RM07/appendixB.html) called for this, so it may be somewhat common, though I don't think the MS OpenType spec requires it.

**Comment by khaledhosny** [comment]
FWIW, `coretext` shaper produces two `.notdef` glyphs, `uniscribe` shaper, however, failed to produce any output.

**Comment by bwrrp** [comment]
Hi, reporter of the Chromium issue here. I've tried mapping U+200B to an actual glyph in the font file, but it appears that HarfBuzz still assigns the space glyph to this code point unless the "preserve default ignorables" flag is passed:

```
hb-shape FontoXMLWhitespace.woff --unicodes=41,200b,41
[.notdef=0+500|space=1+0|.notdef=2+500]

hb-shape FontoXMLWhitespace.woff --unicodes=41,200b,41 --preserve-default-ignorables
[.notdef=0+500|uni200B=1+0|.notdef=2+500]
```

I'm not very familiar with font rendering, but from the docs at https://harfbuzz.github.io/setting-buffer-properties.html it appears that this behavior is intentional? Unfortunately that means that the font side workaround mentioned above does not work, as Chromium does not appear to set this flag when calling the shaper.

**Comment by khaledhosny** [comment]
Font workaround should be possible once #2883 is fixed, by adding a single substitution from the default glyph to an alternate one. This should work with other layout engines as well, since this space replacement behavior is not specific to HarfBuzz.

**Comment by behdad** [comment]
Inspecting outline is doable but not the easiest. It would break with clients using custom fontfuncs.

I think Khaled's proposed solution is better. I'm happy with that for now.

Note that we already have `hb_buffer_set_invisible_glyph()`, which Chrome can use. I expect all heavyweight users of HarfBuzz to use that. So I think it's a non-issue here currently. The space default is for simpler clients.

**Comment by behdad** [comment]
Also note that CoreText seems to return the unreachable gid65535 for this purpose. People might object, but most graphics libraries just work fine if you ask them to render an out-of-bounds gid.

**Comment by jfkthame** [comment]
IIRC, the gid 0xFFFF is explicitly defined as DELETED_GLYPH in AAT,[1] so I'd expect Apple's graphics stack at least to understand this and know to ignore it; I'm not quite as confident about other systems, but it may be fine.

[1] E.g. search for "deleted glyph" in https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6Tables.html

**Comment by behdad** [comment]
> IIRC, the gid 0xFFFF is explicitly defined as DELETED_GLYPH in AAT,[1] so I'd expect Apple's graphics stack at least to understand this and know to ignore it; I'm not quite as confident about other systems, but it may be fine.

Right. I checked with Cairo and it just shows nothing. I thought about it more and it just makes sense. As in: what else would you do?

At any rate; we have `hb_buffer_set_invisible_glyph()` and that's the limit of what we need to do. User can decide what's safe for their graphics library.

**Comment by drott** [comment]
Thanks everyone, I'll roll in the fix for #2883 and not in the Chrome side of this bug that we should employ `hb_buffer_set_invisible_glyph()` - thanks for the pointer.

**Comment by behdad** [comment]
@bwrrp Here's what I suggest you do: map SPACE to am invisible glyph in cmap, then in `ccmp` feature substitute that glyph with the visible one.
